Adding a transactionparser

LimpidCrypto opened the pull request:
Changes marked for commit:
	changed:       xrpl/utils/__init__.py
	new folder:     xrpl/utils/txn_parser

## High Level Overview of Change

Add a transaction parser. It supports "parse_balance_changes", "parse_final_balances", "parse_previous_balances" and "parse_order_book_changes".
Also included the functions in the "xrpl/utils/__init__.py" file.
This is a rewritten parser of the https://github.com/ripple/ripple-lib-extensions/blob/master/transactionparser/src/balancechanges.js in python.

mvadari commented on xrpl/utils/tx_parser.py:
```diff
@@ -0,0 +1,197 @@
+        self.PreviousFields = previousFields
+
+
+def __normalize_node(affectedNode):
```
Internal functions should be prepended by just one `_` not two
```suggestion
def _normalize_node(affectedNode):
```
ditto for the other ones

mvadari commented on xrpl/utils/tx_parser.py:
```diff
@@ -0,0 +1,197 @@
+"""
+
+### Utils ###
+class Normalized_Node():
```
Class names shouldn't have underscores in them
```suggestion
class NormalizedNode():
```

mvadari commented on xrpl/utils/tx_parser.py:
```diff
@@ -0,0 +1,197 @@
+### Utils ###
+class Normalized_Node():
+    def __init__(self, difftype: str, entryType: str, ledgerIndex: str, newFields: dict, finalFields: dict, previousFields: dict) -> None:
+        self.DiffType = difftype
```
```suggestion
        self.diff_type = difftype
```
Please use `snake_case` for variable names.

mvadari commented on xrpl/utils/tx_parser.py:
```diff
@@ -0,0 +1,197 @@
+    return None
+
+
+def __parse_XRP_quantity(node, valueParser):
```
```suggestion
def _parse_xrp_quantity(node, valueParser):
```
Function names shouldn't have capital letters.

mvadari commented:
Here's the PR from xrpl.js for `getBalanceChanges`, for reference

mvadari commented on xrpl/utils/__init__.py:
```diff
@@ -8,6 +8,11 @@
     ripple_time_to_posix,
 )
+from xrpl.utils.tx_parser import (
+    XRPLMetadataExcetion,
```
Typo here - should be `XRPLMetadataException`

mvadari commented on xrpl/utils/tx_parser.py:
```diff
@@ -0,0 +1,277 @@
+import pydash
```
This new package needs to be added to the poetry dependencies. Instructions for how to use that are in `CONTRIBUTING.md`.

mvadari commented on xrpl/utils/tx_parser.py:
```diff
@@ -0,0 +1,277 @@
+from xrpl.utils.xrp_conversions import drops_to_xrp
+
+
+class XRPLMetadataExcetion(XRPLException):
```
```suggestion
class XRPLMetadataException(XRPLException):
```

mvadari commented on xrpl/utils/tx_parser.py:
```diff
@@ -0,0 +1,277 @@
+    return _group_by_address(pydash.compact(pydash.flatten(values)))
+
+
+def parse_balance_changes(metadata: dict) -> Dict[str, List[Dict[str, str]]]:
```
Can you add unit tests for these methods in `tests/unit/utils`?

mvadari commented:
It'll be easier to work with the linter if you set up the pre-commit hook. Instructions for that are in `CONTRIBUTING.md`.

mvadari commented:
You'll need to add the `pydash` package via Poetry, not just editing the `.lock` file.

LimpidCrypto commented:
Thanks for your patience @mvadari . As you may have noticed this is the first time I contribute to such a project :D Also the first time I worked with pyenv and poetry. 
So the unit test works fine now. The only error I still get when I run pre-commit is the following:
```xrpl/utils/tx_parser.py:8: error: Skipping analyzing "pydash": module is installed, but missing library stubs or py.typed marker```

Is this a critical error? As defined here https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports I would have to create a ```py.typed``` file in ```pydash```. Just as in dependencies like e.g. ```websockets```.

mvadari commented:
It's not a critical error - it's not worth dealing with for this PR. 

I think if you can get around using pydash that would be preferable, but if you can't then that's fine.

wojake commented:
When will this be merged? I'd love to use this.

LimpidCrypto commented:
The PR needs to be reviewed first. I also just found a miscalculation in _flip_trustline_perspective. I will fix that tomorrow.

JST5000 commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,231 @@
+"""
+Parse balance changes, final balances and previouse balances of every
```
```suggestion
Parse balance changes, final balances and previous balances of every
```

JST5000 commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,231 @@
+    """All balance changes of the source account (Field: `Account`)
+    which was affected by the transaction.
+    """
+    destination_account_balances: Union[None, Dict[str, List[Dict[str, str]]]] = None
```
nit: Instead of `Union[None, ...]` you can use `Optional[...]`

JST5000 commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,231 @@
+        )
+
+        result: Dict[str, List[Dict[str, str]]] = {}
+        for k, v in parsedQuantities.items():
+            address = k
+            change = v
```
```suggestion
        for address, change in parsedQuantities.items():
```
The values are based on the position, so you can name them to whatever you'd like :)

JST5000 commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,231 @@
+
+        self._parse()
+
+    def _parse(self: Any) -> None:
```
If you'd like to have `self: ParseFinalBalances` instead of `Any`, you just need to put `import __future__` at the top of the file. That allows python to see that you defined `ParseFinalBalances` even if at this point in the file the definition isn't done yet.

LimpidCrypto replied:
Ah, I wondered how I would do that 👍🏼

JST5000 commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,231 @@
+        result: Dict[str, List[Dict[str, str]]] = {}
+        for k, v in parsedQuantities.items():
+            address = k
+            change = v
```
Same change as above

JST5000 commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,231 @@
+            for count, balance in enumerate(balances):
+                final_balances_value = final_balances[account][count]["Value"]
+                final_balances[account][count]["Value"] = str(
+                    Decimal(final_balances_value) - Decimal(balance["Value"])
```
Should this be addition? (Are the changes listed where +100 = Add 100 XRP to the previous value?)

LimpidCrypto replied:
Example:
`BalanceChanges`:
-100 XRP
`FinalBalance`:
900 XRP
`PreviousBalance`:
900 XRP - -100 XRP = 1000 XRP

If the balance change would increase the balance by 100 the calculation of the `PreviousBalance` would be:
900 XRP - 100 XRP = 800 XRP

So subtraction should be correct :)

JST5000 replied:
That makes sense! Thanks for the example :)

JST5000 commented on xrpl/utils/tx_parser/orderbook_changes.py:
```diff
@@ -0,0 +1,74 @@
+    """
+
+    all_orderbook_changes: Dict[str, List[Dict[str, Any]]] = {}
+    """All orderbook_changes"""
```
What does the key represent?

LimpidCrypto replied:
It represents the accounts address :)

JST5000 replied:
Could you add that to the docstring for future reference?

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,847 @@
+
+    Raises:
+        XRPLMetadataException: If field Account is missing.
+        XRPLMetadataException: If field Account is missing.
```
Duplicate line

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,847 @@
+                    value=field_value["value"],
+                )
+            else:  # if 'field_value' type is dict but is no issued currency amount.
+                factured_field_name = factory(field_name)
```
What does factured mean?

LimpidCrypto replied:
I will rename it 👍🏼

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,847 @@
+"""
+
+
+class Balance:
```
This might be the same as an `IssuedCurrencyAmount`. 

You may also want to see if `IssuedCurrency.py` is a useful file for you.

JST5000 replied:
Or maybe not because the `counterparty` doesn't correspond to the currency, it corresponds to the payment?

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,847 @@
+            node.previous_fields.Balance
+        )
+
+    # if value is not 'None' or '0' return 'value' else 'None'.
+    return None if value is None else None if value == 0 else value
```
```suggestion
     if( value and value != 0):
           return value
     else:
          return None
```
There's a lot of None's in this line and it's a bit confusing to read.

LimpidCrypto commented:
Is it my fault the integration tests failed? They failed because the transaction tests did not succeed.

JST5000 commented:
> Is it my fault the integration tests failed? They failed because the transaction tests did not succeed.

That's odd. I'll try to dig into this a bit before the end of today and get back to you.

JST5000 commented:
@LimpidCrypto Turns out it was just a flaky test, sorry about that! (The sequence number didn't match up in the failing test case, and my hypothesis is that for some reason the mock-rippled instance wasn't being restarted, so the sequence number just kept going up. If it happens again I'll debug deeper)

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+
+def parse_order_book_changes(
+    nodes: Iterable[NormalizedNode],
+) -> Any:
```
Small improvement 
```suggestion
) -> List[Any]:
```

JST5000 commented on xrpl/utils/tx_parser/orderbook_changes.py:
```diff
@@ -0,0 +1,78 @@
+        """
+        self.metadata = metadata
+
+    def _define_results(self: Any, result: Dict[str, List[Dict[str, Any]]]) -> None:
```
```suggestion
    def _define_results(self: OrderBookChanges, result: Dict[str, List[Dict[str, Any]]]) -> None:
```

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+    """A standard format for nodes."""
+
+    def __init__(
+        self: Any,
```
```suggestion
        self: NormalizedNode,
```

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+        diffType: str,
+        entryType: str,
+        ledgerIndex: str,
+        newFields: Any,
+        finalFields: Any,
+        previousFields: Any,
```
Are these actually Any? (Or are they a list of some sort?)

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+        The full field state with all its fields as objects.
+    """
+
+    def factory(name: str) -> Any:
```
Would creating a base class you return from factory help cut down on the use of Any? 

(The main reason being that Mypy loses a lot of effectiveness the more `Any`s enter the code because they multiply)

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+
+def _parse_xrp_quantity(
+    node: NormalizedNode,
+    valueParser: Any,
```
If this is a function, you can probably import `Callable` from typing

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+            A list of trust line quantities.
+
+    Returns:
+        Any: A dictionary with all balance changes grouped by addresses.
```
This description implies this could be more specific in typing

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+        node (NormalizedNode): Normalized node.
+
+    Returns:
+        Union[Any, None]:
```
This return type can be more specific (I think it's a string?)

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+    taker_gets: Any,
+    taker_pays: Any,
+    direction: str,
+) -> Any:
```
When you return multiple things you can represent that with `Tuple[first_type, second_type]`

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,850 @@
+        nodes (Iterable[NormalizedNode]): Affected nodes.
+
+    Returns:
+        Any:
```
This can also be a list

LimpidCrypto commented:
## Improve type hints commits:
- Added `normalize_metadata` to format metadata received from subscritions to have one standard format so that typing is easier
- Normalize nodes in the `Parse...` classes so the `metadata` type hint do not need to be repeated that often
- Renamed `Balance` to `AccountBalance` so it won't get in conflict with the `Balance` field in metadata
- Added a lot of type hints
- Formatted some arg/attr names
- Corrected some doc strings

Explanations of added `# type: ignore`:
- `# type: ignore` in `normalize_metadata` is because `mypy` is identifying a false type hint of metadata
- `# type: ignore` every time a function returns a return of `pydash` because `mypy` always identifies them as `Any`

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1054 @@
+                ],
+            ],
+        ],
+    ],
```
Holy crap that's one hell of a type! 

I'd recommend turning this into a class or type alias, that way you can define the type once and then use it everywhere else. 

(That would also let you turn instances of `metadata: Any` above into something like `metadata: Metadata` which is nice and readable :) )

JST5000 replied:
Here's the doc on type aliases - https://mypy.readthedocs.io/en/stable/kinds_of_types.html#type-aliases

LimpidCrypto replied:
I thought that this would be possible but did not find the right solution to that. Thank you I will change that 🙏

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1054 @@
+            str,
+            Union[str, int, Dict[str, Union[str, int, Dict[str, str]]]],
+        ],
+    ]
```
This type is probably a useful sub-type to define for the Metadata type :)

JST5000 commented on xrpl/utils/tx_parser/orderbook_changes.py:
```diff
@@ -0,0 +1,113 @@
+    # all_orderbook_changes = {'account': ['order book changes']}
+    all_orderbook_changes: Dict[str, List[Dict[str, Any]]] = {}
+
+    def __init__(self: OrderBookChanges, metadata: Any) -> None:
```
Does this metadata field re-use the type you defined elsewhere?

LimpidCrypto replied:
I defined the parent classes `metadata` as `Any` because I did not wanted to define the huge type hint once again.
So every child class has the exact type hint but the parent classes don't. This also will be improved by using type aliases :)

JST5000 commented:
For some of the longer types I think it's worth using a type alias even if it's just used once. That'll let you keep the function definitions focused on the implementations instead of the type. (Then if folks want to learn what the type signature actually means they can opt-into that information)

mvadari commented on docs/source/xrpl.utils.tx_parser.rst:
```diff
@@ -0,0 +1,37 @@
+xrpl.utils.tx\_parser package
```
This doc shouldn't be added - the docs for the new function will be in the `xrpl.utils` file

mvadari commented on xrpl/utils/tx_parser/__init__.py:
```diff
@@ -0,0 +1,17 @@
+from xrpl.utils.tx_parser.utils import XRPLMetadataException
+
+__all__ = [
+    "ParseBalanceChanges",
+    "ParseFinalBalances",
+    "ParsePreviousBalances",
+    "ParseOrderBookChanges",
```
All of these methods should use `snake_case`, since they're Python functions

mvadari commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,332 @@
+)
+
+
+class BalanceChanges:
```
Why are these all classes instead of methods? IMO if they're only existing to process the metadata, they should be methods.

LimpidCrypto replied:
That way I think it is easier to access e.g. `.source_account_balances` directly from the class. If you think this is not necessary or too confusing I can easily change them up to functions and just return all balance changes, final balances, ... :)

mvadari commented on xrpl/utils/tx_parser/balance_changes.py:
```diff
@@ -0,0 +1,332 @@
+
+    def __init__(
+        self: ParsePreviousBalances,
+        metadata: Dict[  # metadata received from a tx method
```
You can define this type at the top or in your utils file, which will make it much neater.

mvadari replied:
e.g.
```py
Metadata = Dict[str, ...]
```

mvadari replied:
You can even nest types to make it neater

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+
+def _parse_trustline_quantity(
+    node: NormalizedNode,
+    value_parser: Callable[[NormalizedNode], Union[Decimal, None]],
```
nit:
```suggestion
    value_parser: Callable[[NormalizedNode], Optional[Decimal]],
```

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+def _parse_trustline_quantity(
+    node: NormalizedNode,
+    value_parser: Callable[[NormalizedNode], Union[Decimal, None]],
+) -> Union[List[TrustLineQuantity], None]:
```
nit:
```suggestion
) -> Optional[List[TrustLineQuantity]]:
```

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+
+def parse_quantities(
+    nodes: List[NormalizedNode],
+    value_parser: Callable[[NormalizedNode], Union[Decimal, None]],
```
ditto

JST5000 commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+
+def _calculate_delta(
+    final_amount: Union[AccountBalance, None],
+    previous_amount: Union[AccountBalance, None],
```
ditto

JST5000 reviewed (approved):
Looks good to me - small nitpick about using Optional, but very solid work!

mvadari commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+    pass
+
+
+def is_valid_metadata(
```
```suggestion
def validate_metadata(
```
Using the `is_valid` implies that it's going to be returning a boolean as to whether the metadata is valid or not, but that's not what this method does.

mvadari commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+    if "transaction" in metadata:
+        if "Account" not in metadata["transaction"]:  # type: ignore
+            raise XRPLMetadataException(
+                "Metadata incomplete: Metadata field 'Account' must be included."
```
```suggestion
                "Malformed metadata: Metadata field 'Account' must be included."
```
ditto to the other messages

mvadari commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+        Metadata in standard format.
+    """
+    transaction = metadata["transaction"]
+    assert isinstance(transaction, dict)
```
You can simplify this logic and avoid the `isinstance` if `METADATA_TYPE` and `SUBSCRIPTION_METADATA_TYPE` are defined with `TypedDict`: https://docs.python.org/3/library/typing.html#typing.TypedDict

mvadari commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+    Expiration: Optional[Union[int, str]] = None
+
+
+class NewFields(FieldState):
+    """NewFields"""
+
+    pass
+
+
+class FinalFields(FieldState):
+    """FinalFields"""
+
+    pass
+
+
+class PreviousFields(FieldState):
+    """PreviousFields"""
+
+    pass
```
Why are these defined as separate classes if they're all identical?

mvadari commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+    return normalized_metadata  # type: ignore
+
+
+class FieldName:
+    """FieldName"""
+
+    currency: str
+    issuer: str
+    value: str
+
+
+class FieldState:
+    """FieldState"""
+
+    Balance: Union[AccountBalance, str]
+    LowLimit: Union[AccountBalance, str]
+    HighLimit: Union[AccountBalance, str]
+    Account: str
+    TakerGets: Union[FieldName, str]  # Type: 'NewClass'
+    TakerPays: Union[FieldName, str]  # Type: 'NewClass'
+    Sequence: int
+    Flags: int
+    Expiration: Optional[Union[int, str]] = None
```
These might be easier to work with if you make them dataclasses

mvadari commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+        self.previous_fields = previous_fields
+
+
+def field_factory(
```
This method feels very hacky. Typing doesn't need to be so restrictive in this package that we need to do some metaprogramming in order to use it.

LimpidCrypto replied:
I found a way to make this function way easier 👍

mvadari commented on xrpl/utils/tx_parser/utils.py:
```diff
@@ -0,0 +1,1052 @@
+from xrpl.utils.xrp_conversions import drops_to_xrp
+
+"""
+###########################################
```
This file is very long, and it looks like it has distinct parts. Why is it not split into multiple files?

mvadari reviewed (commented):
Please run `poetry run sphinx-apidoc -o docs/source xrpl` to generate the new documentation files for this PR.

mvadari commented on docs/source/xrpl.utils.txn_parser.rst:
```diff
@@ -0,0 +1,37 @@
+xrpl.utils.txn\_parser package
```
These two docs pages should actually both just contain `:orphan:` - since you're exporting the txn parser stuff at the `xrpl.utils` level, the docs will appear there already, and you don't need separate pages.

Sorry for the confusion.

mvadari commented on tests/unit/utils/test_txn_parser.py:
```diff
@@ -0,0 +1,476 @@
+TEST_TXN = {
+    "engine_result": "tesSUCCESS",
+    "engine_result_code": 0,
+    "engine_result_message": """
```
Why does this have a `.strip()` instead of having the string just be there as is?

mvadari commented on tests/unit/utils/test_txn_parser.py:
```diff
@@ -0,0 +1,476 @@
+                    },
+                    "LedgerEntryType": "RippleState",
+                    "LedgerIndex": """
+0B0388546E94E2B339A52C70CE9B7791FDF104F114C94CEACB2A9440819DD435
```
ditto

mvadari commented on tests/unit/utils/test_txn_parser.py:
```diff
@@ -0,0 +1,455 @@
+TEST_TXN = {
+    "engine_result": "tesSUCCESS",
+    "engine_result_code": 0,
+    "engine_result_message": """The transaction was applied. Only final in a validated ledger.
+""",
```
```suggestion
    "engine_result_message": "The transaction was applied. Only final in a validated ledger.",
```
ditto for the others

mvadari commented on xrpl/utils/txn_parser/balance_changes.py:
```diff
@@ -0,0 +1,137 @@
+            Raw transaction data including the account that
+            sent the transaction and the affected nodes.
+
+    Returns:
+        Dict[str, Any]:
+            All previous balances.
```
nit: this should just be
```suggestion
    Returns:
        All previous balances.
```
ditto with the other docstrings

mvadari commented on xrpl/utils/txn_parser/balance_changes.py:
```diff
@@ -0,0 +1,137 @@
+    by the transaction before it occurred.
+
+    Args:
+        transaction (Union[RawTxnType, SubscriptionRawTxnType]):
```
You don't need to put the types in the docstrings - they'll already be there in the generated documentation

mvadari commented on xrpl/utils/txn_parser/balance_changes.py:
```diff
@@ -0,0 +1,137 @@
+            All final balances.
+    """
+    validate_transaction_fields(transaction_data=transaction)
+    if "transaction" in transaction:
+        transaction = cast(SubscriptionRawTxnType, transaction)
```
You might be able to simplify this to:
```suggestion
    if isinstance(transaction, SubscriptionRawTxnType):
```

LimpidCrypto replied:
Unfortunately that does not work.

mvadari commented on tests/unit/utils/test_txn_parser.py:
```diff
@@ -0,0 +1,454 @@
+        expected = {
+            "rpXhhWmCvDwkzNtRbm7mmD1vZqdfatQNEe": [
+                {
+                    "Counterparty": "rhub8VRN55s94qWKDv6jmDy1pUykJzF3wq",
+                    "Currency": "USD",
+                    "Value": "-0.7042071790427",
+                },
+                {"Counterparty": "", "Currency": "XRP", "Value": "1.000000"},
```
We should keep the standard XRPL parlance, to avoid confusion - so `issuer` instead of `Counterparty`. It would also be more Pythonic/fit the library better if it's all lowercase instead of being capitalized.

mvadari replied:
ditto with all of the other methods as well

LimpidCrypto replied:
If you parse balance changes for example, it returns the balance changes of all parties affected.
For example rPu2feBaViWGmWJhvaF5yLocTVD8FUxd2A bought 8.455426113856 CNY from RippleQK. The balance change for RippleQK looks like:
`'rPT74sUcTBTQhkHVD54WGncoqXEAMYbmH7': [{'Counterparty': 'rafjNPiCrvVniSGSj9UUpAPTvGnHsbeuKn', 'Currency': 'CNY', 'Value': '8.455426113856'}, {'Counterparty': 'rPu2feBaViWGmWJhvaF5yLocTVD8FUxd2A', 'Currency': 'CNY', 'Value': '-8.455426113856'}]}`
The counterparty for RippleQK is rPu2feBaViWGmWJhvaF5yLocTVD8FUxd2A but the account is not the issuer.
I think counterparty would be the better term :) I guess for the order book changes it would make sense.
It should be lowercase.

mvadari commented on xrpl/utils/txn_parser/utils/balance_changes_utils.py:
```diff
@@ -0,0 +1,268 @@
+        Optional[Decimal]:
+            The balance value as Decimal.
+    """
+    if node.new_fields and hasattr(node.new_fields, "Balance"):
```
```suggestion
    if node.new_fields is not None and hasattr(node.new_fields, "Balance"):
```

mvadari commented on xrpl/utils/txn_parser/utils/balance_changes_utils.py:
```diff
@@ -0,0 +1,268 @@
+    if node.new_fields and hasattr(node.new_fields, "Balance"):
+        new_balance = cast(Union[AccountBalance, str], node.new_fields.Balance)
+        return _parse_value(value=new_balance)
+    if node.final_fields and hasattr(node.final_fields, "Balance"):
```
```suggestion
    if node.final_fields is not None and hasattr(node.final_fields, "Balance"):
```

mvadari commented on xrpl/utils/txn_parser/balance_changes.py:
```diff
@@ -0,0 +1,134 @@
+
+    Args:
+        transaction: Raw transaction data including the account that
+            sent the transaction and the affected nodes.
```
```suggestion
            sent the transaction and the metadata.
```

mvadari commented on xrpl/utils/txn_parser/balance_changes.py:
```diff
@@ -0,0 +1,134 @@
+    final_balances = parse_final_balances(transaction=transaction)
+
+    for account, balances in balance_changes.items():
+        for count, balance in enumerate(balances):
+            final_balances_value = final_balances[account][count]["value"]
+            final_balances[account][count]["value"] = str(
+                Decimal(final_balances_value) - Decimal(balance["value"])
+            )
```
```suggestion
        for balance in balances:
            final_balances_value = balances["value"]
            balances["value"] = str(
                Decimal(final_balances_value) - Decimal(balance["value"])
            )
```

mvadari commented:
I know this is pretty far into the process, but could you split up this PR, with one exported method per PR? It would make it much easier to understand what's going on, because it seems like I find something I've previously missed on every review. The PRs can build on top of each other if that's easier.

LimpidCrypto commented:
> I know this is pretty far into the process, but could you split up this PR, with one exported method per PR? It's hard to catch everything on a 2k line PR. It would make it much easier to understand what's going on, because it seems like I find something I've previously missed on every review, and that just keeps dragging this out. The PRs can build on top of each other if that's easier.

@mvadari I can close this pull request then, right?

mvadari commented:
> @mvadari I can close this pull request then, right?

Yes

LimpidCrypto commented:
The four exported functions will gradually get their own PR
